contracts/TapETH.sol

```diff
@@ -0,0 +1,529 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.18;
+import "@openzeppelin/contracts/utils/math/Math.sol";
+import "./interfaces/ITapETH.sol";
+
+error InsufficientAllowance(uint256 currentAllowance, uint256 amount);
+error InsufficientBalance(uint256 currentBalance, uint256 amount);
+
+/**
+ * @title Interest-bearing ERC20-like token for Tapio protocol
+ * @author Nuts Finance Developer
+ * @notice ERC20 token minted by the StableSwap pools.
+ * @dev TapETH is ERC20 rebase token minted by StableSwap pools for liquidity providers.
+ * TapETH balances are dynamic and represent the holder's share in the total amount
+ * of tapETH controlled by the protocol. Account shares aren't normalized, so the
+ * contract also stores the sum of all shares to calculate each account's token balance
+ * which equals to:
+ *
+ *   shares[account] * _getTotalPooledEther() / _getTotalShares()
+ * where the _getTotalPooledEther() returns the total supply of tapETH controlled by the protocol.
+ */
+
+contract TapETH is ITapETH {
+  using Math for uint256;
+  uint256 internal constant INFINITE_ALLOWANCE = ~uint256(0);
+
+  uint256 private totalShares;
+  uint256 private _totalSupply;
+  uint256 private totalRewards;
+  uint256 public buffer;
```

Comment:
from today's discussion, we need more work on the `buffer` idea. Will take this out as part of the testnet deployment.